File: haskell/src/Euler/Problem/Problem11.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE OverloadedStrings #-}

module Euler.Problem.Problem11 (answer) where

import Data.FileEmbed     ( embedFile )
import Data.Text          ( Text )
import Data.Text.Encoding ( decodeUtf8 )

import qualified Data.Either   as Either
import qualified Data.Text      as Text
import qualified Data.Text.Read as Read

import Euler.Util.List    ( sliding )

answer :: Integer
answer = maximum $ map product groups
  where

    groups = foldMap (sliding 4) (concat grids)

    grids = [ grid
            , transpose grid
            , shift grid
            , (shift . reverseRows) grid ]

    transpose [] = []
    transpose rows
      | length (head rows) == 0 = []
      | otherwise = (map head rows) : transpose (map tail rows)

    shift rows = transpose $ map shiftRow $ zip rows [0..]
      where shiftRow (row, i) = (replicate i 0) ++ row ++
                                (replicate (2 * (length row)) 0)

    reverseRows rows = map reverse rows

grid :: [[Integer]]
grid = map parseLine $ Text.lines inputText
  where parseLine = map fst . Either.rights . (map Read.decimal) . Text.words

inputText :: Text
inputText = decodeUtf8 $(embedFile "../problems/11-data.txt")

File: scala/src/main/scala/euler/problem/problem11.scala
package euler.problem

package object problem11 {

  type Grid = Array[Array[Int]]

  def shifted(grid: Grid): Grid =
    grid.zipWithIndex.map({ case (row: Array[Int], i: Int) =>
      (Array.fill(i)(0) ++ row).padTo(2 * grid.length, 0)
    }).transpose[Int]

  def answer(input: String): Int = {

    val grid: Grid =
      input.split("\n").map(_.trim).filter(_.nonEmpty)
        .map(_.split(" ").map(_.toInt))

    Stream(
      grid,
      grid.transpose[Int],
      shifted(grid),
      shifted(grid.map(_.reverse))
    )
    .flatten.view.flatMap(_.sliding(4).map(_.product)).max
  }

}
